modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,295 @@
+
+	/**
+	 * Job ID.
+	 *
+	 * @since n.e.x.t
+	 * @var int
+	 */
+	private $id;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $job_id Job ID. Technically this is the ID of a term in the 'background_job' taxonomy.
+	 */
+	public function __construct( $job_id ) {
+		$this->id = absint( $job_id );
+	}
+
+	/**
+	 * Returns the id for the job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int Job ID. Technically this is a term id for `background_job` taxonomy.
+	 */
+	public function get_id() {
+		return $this->id;
+	}
+
+	/**
+	 * Retrieves the job data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Job data.
+	 */
+	public function get_data() {
+		return (array) get_term_meta( $this->id, self::META_KEY_JOB_DATA, true );
+	}
+
+	/**
+	 * Retrieves the job action.
+	 *
+	 * This action will be used in custom action triggered by background process
+	 * runner. Action will be like `perflab_job_{action}`.
+	 * Consumer code can hook onto this action to perform necessary task for the job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job action.
+	 */
+	public function get_action() {
+		$job = get_term( $this->id, PERFLAB_BACKGROUND_JOB_TAXONOMY_SLUG );
+		return $job->name;
+	}
+
+	/**
+	 * Determines whether a job should run for current request or not.
+	 *
+	 * It determines by checking if job is already running or completed.
+	 * It also checks if max number of attempts have been attempted for the job.
+	 * For any such condition, it will return false, else true.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool Flag to tell whether this job should run or not.
+	 */
+	public function should_run() {
+		// If job doesn't exist or completed, return false.
+		if ( ! $this->exists() || $this->is_completed() ) {
+			return false;
+		}
+
+		if ( $this->is_running() ) {
+			return false;
+		}
+
+		/**
+		 * Number of attempts to try to run a job.
+		 *
+		 * Repeated attempts may be required to run a failed job. Default 3 attempts.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param int $attempts Number of attempts allowed for a job to run. Default 3.
+		 */
+		$max_attempts = (int) apply_filters( 'perflab_job_max_attempts_allowed', 3 );
+
+		// If number of attempts have been exhausted, return false.
+		if ( $this->get_attempts() >= $max_attempts ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Marks the job as completed.
+	 *
+	 * 1. Mark the job status as completed.
+	 * 2. It will also save the timestamp (in seconds) at which the job was completed.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function complete() {
+		$this->set_status( self::JOB_STATUS_COMPLETE );
+		// If job is complete, set the timestamp at which it was completed.
+		update_term_meta( $this->id, self::META_KEY_JOB_COMPLETED_AT, time() );
+	}
+
+	/**
+	 * Marks the job as queued.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function queue() {
+		$this->set_status( self::JOB_STATUS_QUEUED );
+	}
+
+	/**
+	 * Set the start time of job.
```

Comment:
Function/method descriptions should always start with a third-person verb.
```suggestion
	 * Sets the start time of job.
```